\documentclass{article}
\input{../config/commontex}

\title{Release notes}
\author{Lobry, J.R. \and Nec\c{s}ulea, A. \and Palmeira, L. \and Penel, S.}

\begin{document}
\SweaveInput{../config/commonrnw.rnw}
\maketitle
% BEGIN - DO NOT REMOVE THIS LINE

\section*{1.1 series}

\subsection*{release 1.1-3}

\begin{itemize}

\item Function \texttt{read.alignment()} has gained a new argument
  \texttt{forceToLower} defaulting to TRUE to force lower case in
  the character of the sequence (this is for a smoother interaction
  with the package \texttt{ape}). The argument \texttt{File} is now
  deprecated and a warning is issued when used instead of \texttt{file}. 
  The example in the function \texttt{kaks()} has been corrected
  to avoid this warning when reading the example files.

\item New low level utility function \texttt{knowndbs()} and its
  shortcut \texttt{kdb()} to get a description of databases 
  known by an ACNUC server.

\item New low level utility function \texttt{autosocket()} to get the
  socket connection to the last opened ACNUC database.

\item New function \texttt{countsubseqs()} to get the number of 
  subsequences in an ACNUC list.

\item New function \texttt{savelist()} to save sequence names or 
  accession numbers from an ACNUC list into a local file.

\item New function \texttt{ghelp()} to get help from an ACNUC server.

\item New function \texttt{modifylist()} to modify a previously existing 
  ACNUC list by selecting sequences either by length, either by date, 
  either for the presence of a given string in annotations.

\item New low level function \texttt{getlistate()} to Ask for 
  information about an ACNUC list.

\item New low level function \texttt{setlistname()} to set the name of 
  a list from an ACNUC server.

\item New function \texttt{residuecount()} to count the total number 
  of residues (nucleotides or aminoacids) in all sequences of an ACNUC
  list of specified rank.

\item New function \texttt{isenum()} and its shortcut \texttt{isn()}
  to get the ACNUC number of a sequence from its name or accession
  number.

\item New function \texttt{prettyseq()} to get a text representation 
  of a sequence from an ACNUC server.

\item New function \texttt{gfrag()} to extract sequence identified by 
  name or by number from an ACNUC server.

\item The details of the socket connection are no more stored in
  the slot \texttt{socket} for objects of class \texttt{seqAcnucWeb}:
  this slot is now deleted. As a consequence, the argument
  \texttt{socket} in function \texttt{as.SeqAcnucWeb()} has been
  removed and there is now a new
  argument \texttt{socket = "auto"} in functions \texttt{getAnnot()},
  \texttt{getFrag()}, \texttt{geyKeyword()}, \texttt{getLocation()},
  and \texttt{getSequence()}. The default value \texttt{"auto"} means
  that the details of the socket connection are taken automatically
  when necessary from the last opened bank. The size of local lists
  of sequences is reduced by about a third now as compared to the
  previous version.

\item New function \texttt{print.seqAcnucWeb()} to print objects from
  class \texttt{seqAcnucWeb}.

\item Internal function \texttt{parser.socket()} has been optimized and
  is about four times faster now. This decreases the time needed by the
  \texttt{query()} function.

\end{itemize}

\subsection*{release 1.1-2}


\begin{itemize}

\item New function \texttt{trimSpace()} to remove leading and trailing
  spaces in string vectors.

\item Function \texttt{splitseq()} is no more based on \texttt{substring()},
  it is now more efficient for long sequences.

\item A sanity check test was added in the documentation file
  for the function \texttt{syncodons()}.

\item The way this manual is produced is now documented in the
  \texttt{doc/src/template/} folder.

\item A bug in function \texttt{oriloc()} was reported on 23 Jul 2007
  by Michael Kube: using directly genBank files was no more possible.
  The culprit was \texttt{gbk2g2()} that turns genBank files into
  glimmer files version 2 when \texttt{oriloc()} default is to use
  version 3 files. The \texttt{glimmer.version} argument is now forced
  to 2 when working with genBank files to fix this problem.
 
\item Function \texttt{zscore()} has now a new argument
  \texttt{exact} (which is only effective for the option
  \texttt{model = base}). This argument, when set to 
  \texttt{TRUE} allows for the exact analytical computation 
  of the zscore under this model, instead of the approximation for
  large sequences. It is set to \texttt{FALSE} by
  default for backward compatibility.
\end{itemize}

\subsection*{release 1.1-1}


\begin{itemize}
\item A bug was reported by Sylvain Mousset on 14 Jul 2007 in function
\texttt{dist.alignment()}: when called with sequences in lower case
letters, some sequences were modified. This should no more be the case:
<<bugSylvain,fig=F>>=
ali <- list(nb=4, nam=c("speciesA", "speciesB", "speciesC", "speciesD"),
seq=c("ACGT","acgt","ACGT","ACGT"))
class(ali) <- "alignment"
print(ali$seq)
print(dist.alignment(ali))
print(ali$seq)
@ 

\item The CITATION file has been updated so that now \texttt{citation("seqinr")} returns the full complete reference for the package seqinR.

\item Non ASCII characters in documentation
(*.Rd) files have been removed. Declaration of the encoding as 
latin1 when necessary is now present. 
The updated documentation files are: \texttt{
dinucl.Rd,
gb2fasta.Rd,
get.ncbi.Rd,
lseqinr.Rd,
n2s.Rd,
prochlo.Rd,
s2c.Rd,
SeqAcnucWeb.Rd,
SeqFrag.Rd,
toyaa.Rd,
words.pos.Rd,
words.Rd,
zscore.Rd}.

\item Function \texttt{GC()} and by propagation functions \texttt{GC1()},
\texttt{GC2()} and \texttt{GC3()} have gained a new argument \texttt{oldGC}
allowing to compute the G+C content as in releases up to 1.0-6 included.
The code has been also modified to avoid divisions by zero with very
small sequences.

\item New function \texttt{rot13()} that returns the ROT-13 encoding of
a string of characters.

\end{itemize}

\section*{1.0 series}

\subsection*{release 1.0-7}

\begin{itemize}

\item A new \emph{experimental} function \texttt{extractseqs()} to download
sequences thru zlib compressed sockets from an ACNUC server is released. 
Preliminary tests suggest that working with about 100,000 CDS is possible with 
a home ADSL connection. See the manual for some \texttt{system.time()}
examples.

\item As pointed by e-mail on 16 Nov 2006 by Emmanuel Prestat the URL
used in \texttt{dia.bactgensize()} was no more available, this has been fixed
in the current version.

\item As pointed by e-mail on 16 Nov 2006 by Guy Perri{\`e}re, the
function \texttt{oriloc()} was no more compatible with glimmer\footnote{
Glimmer is a program to predict coding sequences in microbial genomes \cite{SalzbergSL1998, DelcherAL1999}.
} 3.0
outputs. The function has gained a new argument \texttt{glimmer.version}
defaulting to 3, but the value 2 is still functional for backward compatibility
with old glimmer outputs.

\item As pointed by e-mail on 24 Oct 2006 by Lionel Guy 
(\url{http://pbil.univ-lyon1.fr/seqinr/seqinrhtmlannuel/03/0089.html})
there was no default value for the \texttt{as.string} argument in
the \texttt{getSequence.SeqFastadna()}. A default \texttt{FALSE}
value is now present for backward compatibility with older code.

\item New utility vectorized function \texttt{stresc()}  to escape \LaTeX~special 
characters present in a string.

\item New low level function \texttt{readsmj()} available.

\item A new function \texttt{readfirstrec()} to get 
the record count of the specified ACNUC index file is now available.

\item Function \texttt{getType()} called without arguments will now use
the default ACNUC database to return available subsequence types.

\item Function \texttt{read.alignment()} now also accepts \texttt{file} in
addition to \texttt{File} as argument.

\item A new function \texttt{rearranged.oriloc()} is available. This
method, based on \texttt{oriloc()}, can be used to detect the effect of
the replication mechanism on DNA base composition asymmetry, in
prokaryotic chromosomes.

\item New function \texttt{extract.breakpoints()}, used to extract
breakpoints in rearranged nucleotide skews. This function uses the
\texttt{segmented} package to define the position of the breakpoints.

\item New function \texttt{draw.rearranged.oriloc()} available, to plot
nucleotide skews on artificially rearranged prokaryotic chromosomes.

\item New function \texttt{gbk2g2.euk()} available. Similarly to 
\texttt{gbk2g2()}, this function extracts the coding sequence annotations
from a GenBank format file. This function is specifically designed for
eukaryotic sequences, \textit{i.e.} with introns. The output file will contain
the coordinates of the exons, along with the name of the CDS to which
they belong.

\item After an e-mail by Marcelo Bertalan on 26 Mar 2007, a bug in
\texttt{oriloc()} when the \texttt{gbk} argument was \texttt{NULL}
was found and fixed by Anamaria Nec\c{s}ulea.

\item Functions \texttt{translate()} and \texttt{getTrans()} have gained
a new argument \texttt{NAstring} to represent untranslatable amino-
acids, defaulting to character "X".

\item There was a typo for the total number of printed bases in
the ACNUC books \cite{GautierC1982a, GautierC1982b} : 474,439 should be 526,506.

\item Function \texttt{invers()} has been deleted.

\item Functions \texttt{translate()}, \texttt{getTrans()} and \texttt{comp()} have gained a 
new argument \texttt{ambiguous} defaulting to FALSE allowing to handle 
ambiguous bases. If TRUE, ambiguous bases are taken into account so that 
for instance GGN is translated to Gly in the standard genetic code.

\item New function \texttt{amb()} to return the list of nucleotide matching 
a given IUPAC nucleotide symbol.

\item Function \texttt{count()} has gained a new argument \texttt{alphabet}
so that oligopeptides counts are now possible. Thanks to Gabriel Valiente
for this suggestion. The functions \texttt{zscore()}, \texttt{rho()} and
\texttt{summary.SeqFastadna()} have also an argument \texttt{alphabet} which
is forwarded to \texttt{count()}.

\end{itemize}

\subsection*{release 1.0-6}

Release 1.0-6 is a minor release to fix a problem found and solved by Kurt Hornik
(namely a change from \texttt{SET\_ELEMENT} to \texttt{SET\_STRING\_ELT}
in C code for \texttt{s2c()} in file \texttt{util.c}). The few changes are
as follows.

\begin{itemize}

\item More typographical option for the output \LaTeX~table of \texttt{tablecode()}
are now available to outline deviations from the standard genetic code (see example in the
appendix "genetic codes" of the manual).

\item A new dataset \texttt{aaindex} extracted from the aaindex database
\cite{aaindex1, aaindex2, aaindex3} is now available. It contains a list
of 544 physicochemical and biological properties for the 20 amino-acids

\item The default value for argument \texttt{dia} is now \texttt{FALSE}
in function \texttt{tablecode()}.

\item The example code for \texttt{data(chargaff)} has been changed.

\end{itemize}

\subsection*{release 1.0-5}

\begin{itemize}

\item A  new function \texttt{dotPlot()} is now available.

\item A new function \texttt{crelistfromclientdata()} is now available to
create a list on the server from a local file of sequence names, sequence
accession numbers, species names, or keywords names.

\item A new function \texttt{pmw()} to compute the molecular weight of
a protein is now available.

\item A new function \texttt{reverse.align()} contributed by Anamaria Nec\c{s}ulea
is now available to align CDS at the protein level and then reverse translate this at
the nucleic acid level from a \texttt{clustalw} output. This can be done on the fly
if \texttt{clustalw} is available on your platform.

\item An undocumented behavior was reported by Guy Perri{\`e}re for \texttt{uco()}
when computing RSCU on sequences where an amino-acid is missing. There is
now a new argument \texttt{NA.rscu} that allows the user to force the
missing values to his favorite magic value.

\item There was a bug in \texttt{read.fasta()}: some sequence names were
truncated, this is now fixed (thanks to Marcus G. Daniels for pointing this).
In order to be more consistent with standard functions such as \texttt{read.table()}
or \texttt{scan()}, the file argument starts now with a lower case letter (\texttt{file})
in function \texttt{read.fasta()}, but the old-style \texttt{File} is still
functional for forward-compatibility. There is a new logical argument in \texttt{read.fasta()}
named \texttt{as.string} to allow sequences to be returned as strings instead of
vector of single characters. The automatic conversion of DNA sequences into
lower case letters can now be disabled with the new logical argument
\texttt{forceDNAtolower}. It is also possible to disable the automatic attributes
settings with the new logical argument \texttt{set.attributes}.

\item A new function \texttt{write.fasta()} is now available.

\item The function \texttt{kaks()} now forces character in sequences to upper case.
This default behavior can be neutralized in order to save time by setting the 
argument \texttt{forceUpperCase} to \texttt{FALSE}.

\end{itemize}

\subsection*{release 1.0-4}

\begin{itemize}
\item The scaling factor $n_{\bullet\bullet}$ was missing in equation \ref{chi}.
\item The files \texttt{louse.fasta}, \texttt{louse.names}, \texttt{gopher.fasta}, \texttt{gopher.names}
and \texttt{ortho.fasta} that were used for examples in the previous version of this document are
no more downloaded from the internet since they are now distributed in the \texttt{sequences/} folder
of the package.
\item An example of synonymous and non synonymous codon usage analysis was added to the
vignette along with two toy data sets (\texttt{toyaa} and \texttt{toycodon}).
\item A FAQ section was added to the vignette.
\item A bug in \texttt{getAnnot()} when the number of lines was zero is now fixed.
\item There is now a new argument, \texttt{latexfile}, in \texttt{tablecode()} to export genetic codes
tables in a \LaTeX~document, for instance table \ref{../tables/code3.tex} and table \ref{../tables/code4.tex} here.
\item There is now a new argument, \texttt{freq}, in \texttt{count()}
  to compute word frequencies instead of counts.
\item Function \texttt{splitseq()} has been entirely rewritten to improve speed.
\item Functions computing the G+C content: \texttt{GC(), GC1(), GC2(),
  GC3()} were rewritten to improve speed, and their document files
  were merged to facilitate usage.
\item The following new functions have been added:
\begin{itemize}
\item \texttt{syncodons()} returns all synonymous codons for a given
  codon. Argument \texttt{numcode} specifies the desired genetic code.
\item \texttt{ucoweight()} returns codon usage bias on a sequence as
  the number of synonymous codons present in the sequence for each
  amino acid.
\item \texttt{synsequence()} generates a random coding sequence which
      is synonymous to a given sequence and has a chosen codon usage
      bias.
\item \texttt{permutation()} generates a new sequence from a given
  sequence, while maintaining some constraints from the given sequence
  such as nucleotide frequency, codon usage bias, ...
\item \texttt{rho()} computes the rho statistic on dinucleotides as
  defined in \cite{Karlin}.
\item \texttt{zscore()} computes the zscore statistic on dinucleotides
  as defined in \cite{UV}.
\end{itemize}
\item Two datasets (\texttt{dinucl} and \texttt{prochlo}) were added
  to illustrate these new functions.

\end{itemize}

\subsection*{release 1.0-3}

\begin{itemize}
\item The new package maintainer is Dr. Simon Penel, PhD, who has now a fixed position in the
laboratory that issued \seqinr~(\texttt{penel@biomserv.univ-lyon1.fr}). Delphine Charif was
successful too to get a fixed position in the same lab, with now a different research task (but who knows?).
Thanks to the close vicinity of our pioneering maintainers the transition was sweet. The DESCRIPTION
file of the \seqinr{} package has been updated to take this into account.

\item The reference paper for the package is now \textit{in press}. We do not have the full
reference for now, you may use \texttt{citation("seqinr")} to check if it is complete now:
<<cite, fig=F, eval=T>>=
citation("seqinr")
@

\item There was a bug when sending a \texttt{gfrag} request to the server for long (Mb range) 
sequences. The length argument was converted to scientific notations that are not understand by the
server. This is now corrected and should work up the the Gb scale.

\item The \texttt{query()} function has been improved by de-looping list element info request,
there are now download at once which is much more efficient. For example, a query from a
researcher-home ADSL connection with a list with about 1000 elements was 60 seconds and
is now only 4 seconds (\textit{i.e.} 15 times faster now).

\item A new parameter \texttt{virtual} has been added to \texttt{query()} 
so that long lists can stay on the server without trying to download
them automatically. A query like \texttt{query(s\$socket,"allcds","t=cds", virtual = TRUE)} is 
now possible.

\item Relevant genetic codes and frames are now automatically propagated.

\item \Seqinr{}~sends now its name and version number to the server.

\item Strict control on ambiguous DNA base alphabet has been relaxed.

\item Default value for parameter \texttt{invisible} of function \texttt{query()} is now \texttt{TRUE}.

\end{itemize}


\SweaveInput{../config/sessionInfo.rnw}

% END - DO NOT REMOVE THIS LINE


%%%%%%%%%%%%  BIBLIOGRAPHY %%%%%%%%%%%%%%%%%
\clearpage
\addcontentsline{toc}{section}{References}
\bibliographystyle{plain}
\bibliography{../config/book}
\end{document}
